src/main/java/org/openmaptiles/layers/Poi.java

```diff
@@ -125,19 +153,140 @@ private int minzoom(String subclass, String mappingKey) {
+    // adjusted formula from `Utils.getMinZoomForLength()`, given that 1/10 does not match `1 / (2^<something>)`
+    double zoom = -(Math.log(oneSideWorld * SQRT10) / LOG2);
+
+    // Say Z13.01 means bellow threshold, Z13.00 is exactly threshold, Z12.99 is over threshold,
+    // hence Z13.01 and Z13.00 will be rounded to Z14 and Z12.99 to Z13 (e.g. `floor() + 1`).
+    // And to accommodate for some precision errors (observed for Z9-Z11) we do also `- 0.1e-10`.
+    int result = (int) Math.floor(zoom - 0.1e-10) + 1;
+
+    return Math.min(14, Math.max(10, result));
+  }
+
+  @Override
+  public void release() {
+    aggStops.clear();
+  }
+
   @Override
   public void process(Tables.OsmPoiPoint element, FeatureCollector features) {
-    // TODO handle uic_ref => agg_stop
-    setupPoiFeature(element, features.point(LAYER_NAME));
+    if (element.uicRef() != null && AGG_STOP_SUBCLASS_ORDER.containsKey(element.subclass())) {
+      // multiple threads may update this concurrently
+      synchronized (this) {
+        aggStops.computeIfAbsent(element.uicRef(), key -> new ArrayList<>()).add(element);
+      }
+    } else {
+      setupPoiFeature(element, features.point(LAYER_NAME), null);
+    }
+  }
+
+  private void processAggStop(Tables.OsmPoiPoint element, FeatureCollector.Factory featureCollectors,
+    Consumer<FeatureCollector.Feature> emit, Integer aggStop) {
+    try {
+      var features = featureCollectors.get(SimpleFeature.fromWorldGeometry(element.source().worldGeometry()));
+      setupPoiFeature(element, features.point(LAYER_NAME), aggStop);
+      for (var feature : features) {
+        emit.accept(feature);
+      }
+    } catch (GeometryException e) {
+      e.log(stats, "agg_stop_geometry_2",
+        "Error getting geometry for the stop " + element.source().id() + " (agg_stop)");
+    }
+  }
+
+  /**
+   * We've put aside some stops for {@code agg_stop} processing and we do that processing here.
+   * <p>
+   * The main point is to group together stops with same {@code uid_ref} and then order them first based on subclass
+   * (see {@code AGG_STOP_ORDER}) and then based on distance from centroid (calculated from all the stops). The first
+   * one gets {@code agg_stop=1}, the rest will be "normal" (e.g. no {@code agg_stop} attribute).
+   * <p>
+   * ref: <a href=
+   * "https://github.com/openmaptiles/openmaptiles/blob/master/layers/poi/poi_stop_agg.sql#L26,L28">poi_stop_agg.sql</a>
+   */
+  @Override
+  public void finish(String sourceName, FeatureCollector.Factory featureCollectors,
+    Consumer<FeatureCollector.Feature> emit) {
+    if (OpenMapTilesProfile.OSM_SOURCE.equals(sourceName)) {
+      var timer = stats.startStage("agg_stop");
+      LOGGER.info("Processing {} agg_stop sets", aggStops.size());
+
+      // possible TODO: run in parallel?
+      for (var aggStopSet : aggStops.values()) {
+        if (aggStopSet.size() == 1) {
+          processAggStop(aggStopSet.get(0), featureCollectors, emit, 1);
```

Comment:
Now that `agg_stop` is implemented, can we remove the line in the README that says it's a discrepancy between planetiler-openmaptiles and openmaptiles?